### PR TITLE
20201118 token expiration

### PR DESCRIPTION
**Pull request by peterthomassen**
This PR is based on work from another branch that is reviewed in #474. Only the last 5 commits are relevant here.

### CODE REVIEW

**Review by nils-wisiol — APPROVED** [review]
lgtm, one usability comment:

I find it unintuitive that in the webapp token table, when I click the "Can manage tokens" slider, I still have to save that row. From my experience with Android, I'd assume it saves automatically/instantly. The background color change didn't help much in my case, as I only had two rows: the login token, dark grey becase disabled, and the current row. The current row was white to begin with, then light grey on hover, then later yellow because of unsaved changes - which I didn't notice. Two ways to improve this could be (in order of my preference): 1. introduce an indicator of unsaved changes on top of the table where the blue bar is currently (preferably in a way that doesn't move the page when it appears/disappears). 2. hide the save button of each row unless there are changes to save 3. auto-save the changes on the slider 4. popup-warning when you leave the page with unsaved changes

As those are more general concerns with how the webapp tables work, we can do this as part of this PR or later, in which case I'd like to ask you to create an issue for it :)

File: api/desecapi/authentication.py [hunk 1]
```diff
@@ -17,12 +17,20 @@
...
-        except TypeError:  # TypeError: cannot unpack non-iterable NoneType object
+            user, token = super().authenticate(request)  # may raise exceptions.AuthenticationFailed if token is invalid
+        except TypeError:  # if no token was given
             return None  # unauthenticated
 
+        if not token.is_valid:
+            raise exceptions.AuthenticationFailed('Invalid token.')
+
+        token.last_used = timezone.now()
+        token.save()
```
[the remark this answers is not in the document]
Comment (nils-wisiol) [reply api/desecapi/authentication.py:32]: This is not behavior introduced by this PR. Still, question is still worth thinking about it

[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/authentication.py:32]: I don't think so. First, I believe that this does not happen inside a transaction; second, even if it were, one requests would simply delay the other, but not raise an exception (if I got it right).

**Comment by peterthomassen** [comment]
I opened the issue regarding the usability comment (#485).

File: docs/auth/tokens.rst [hunk 2]
```diff
@@ -72,6 +81,29 @@ Field details:
...
+``max_age``
+    :Access mode: read, write
+    :Type: string (time duration: ``[DD] [HH:[MM:]]ss[.uuuuuu]``) or ``null``
+
+    Maximum token age.  If ``created + max_age`` is less than the current time,
+    the token is invalidated.  Invalidated tokens are not automatically deleted
+    and can be resurrected by adjusting the expiration settings (using another
+    valid token with sufficient privileges).
+
+    If ``null``, the setting is disabled.
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply docs/auth/tokens.rst:93]: done

File: docs/auth/tokens.rst [hunk 3]
```diff
@@ -72,6 +81,29 @@ Field details:
...
+    :Access mode: read, write
+    :Type: string (time duration: ``[DD] [HH:[MM:]]ss[.uuuuuu]``) or ``null``
+
+    Maximum allowed time period of disuse without invalidating the token.  If
+    ``max(created, last_used) + max_unused_period`` is less than the current
+    time, the token is invalidated.  Invalidated tokens are not automatically
+    deleted and can be resurrected by adjusting the expiration settings (using
+    another valid token with sufficient privileges).
+
+    If ``null``, the setting is disabled.
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply docs/auth/tokens.rst:105]: done

File: docs/auth/tokens.rst [hunk 4]
```diff
@@ -146,6 +178,15 @@ configuration during creation:
 - ``perm_manage_tokens``:  If set to ``true``, the token can be used to
   authorize token management operations (as described in this chapter).
 
+Additionally, you can configure an expiration policy with the following fields:
+
+- ``max_age``:  Force token expiration when a certain time period has passed
+  since its creation.  If ``null``, the rule is disabled.
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply docs/auth/tokens.rst:184]: done

File: docs/auth/tokens.rst [hunk 5]
```diff
@@ -146,6 +178,15 @@ configuration during creation:
...
+
+- ``max_unused_period``:  Require that the token is used a least once within
+  the given time period to prevent it from expiring.  If ``null``, the rule is
+  disabled.
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply docs/auth/tokens.rst:188]: done

**Pull request merged**